src/mongoc/mongoc-topology-description.c

```diff
@@ -0,0 +1,996 @@
+                                                           mongoc_server_description_t   *server)
+{
+   _mongoc_topology_description_set_state(topology, MONGOC_TOPOLOGY_SHARDED);
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_topology_description_transition_unknown_to_rs_no_primary --
+ *
+ *       Encapsulates transition from cluster state UNKNOWN to
+ *       RS_NO_PRIMARY. Sets the type to RS_NO_PRIMARY,
+ *       then updates the replica set accordingly.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Changes topology state.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_transition_unknown_to_rs_no_primary (mongoc_topology_description_t *topology,
+                                                                  mongoc_server_description_t   *server)
+{
+   _mongoc_topology_description_set_state(topology, MONGOC_TOPOLOGY_RS_NO_PRIMARY);
+   _mongoc_topology_description_update_rs_without_primary(topology, server);
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_topology_description_remove_and_check_primary --
+ *
+ *       Remove this server from being monitored, then check that the
+ *       current topology has a primary.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Stop monitoring server.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_remove_and_check_primary (mongoc_topology_description_t *topology,
+                                                       mongoc_server_description_t   *server)
+{
+   _mongoc_topology_description_remove_server(topology, server);
+   _mongoc_topology_description_check_if_has_primary(topology, server);
+}
+
+/*
+ *--------------------------------------------------------------------------
+ *
+ * _mongoc_topology_description_update_unknown_with_standalone --
+ *
+ *       If the cluster doesn't contain this server, do nothing.
+ *       Otherwise, if the topology only has one seed, change its
+ *       type to SINGLE. If the topology has multiple seeds, it does not
+ *       include us, so remove this server and stop monitoring us.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       Changes the topology type, might remove server from monitor.
+ *
+ *--------------------------------------------------------------------------
+ */
+static void
+_mongoc_topology_description_update_unknown_with_standalone (mongoc_topology_description_t *topology,
+                                                             mongoc_server_description_t   *server)
+{
+   if (!_mongoc_topology_description_has_server(topology, server->connection_address)) return;
+
+   if (topology->servers->items_len > 1) {
+      /* This cluster contains other servers, it cannot be a standalone. */
```

Comment:
nit: even though this logic is correct it would be nice if it directly mirrored the [SDAM pseudocode](http://emptysqua.re/server-discovery-and-monitoring.html#actions)